### ISSUE TITLE
Unexpected thread usage increase

### ISSUE THREAD

**Issue by nrcventura** [newrelic/newrelic-dotnet-agent#119]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

**Description**
Something is causing web applications instrumented by the agent to periodically use roughly 500 threads instead of around 50-60 threads.

**Expected Behavior**
We expect thread usage to increase when the agent is running because of the native threads that are required for certain .NET profiling API calls, metric sampling (CPU, Memory, Garbage Collection, Thread Info), sending data to New Relic, and continuation based async timing, but it should not increase the thread count by a large amount.

**[Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results**
See this [internal link](https://newrelic.atlassian.net/browse/GTSE-2112) for some of the troubleshooting that has been done previously to try discover what's causing the spike in thread usage. The link details which agent features and instrumentation that were disabled, as well as the agent settings that were changed, to try to minimize the increase in thread usage.

**Steps to Reproduce**
N/A

**Your Environment**
N/A

**Additional context**
I've seen this thread explosion from time to time with some test applications during performance testing. However, I do not yet understand what is exactly triggering this explosion. Is it just a combination of circumstances such as: 1. A certain amount of load within the application, 2. A certain amount of load on the system running the application, 3. Samplers, Harvests, transactions completing, and async continuations needing to execute around the same time, 4. Is this happening after a bunch of work was delayed due to a blocking GC?

**Comment by angelatan2** [comment]
12/10/2021 - Let's do a sanity check first on this problem.

**Comment by JcolemanNR** [comment]
Based on my testing during Firefly, I was not able to see an increase as described. 

I have left a test environment running with the infrastructure agent as described in this ticket and have a meeting scheduled with @nrcventura to review the results tomorrow before closing.

**Comment by JcolemanNR** [comment]
Here is the application under load:
https://staging.newrelic.com/accounts/273070/applications/18422

Here is the infrastructure query to show thread count for the process being instrumented/stressed:
https://staging.onenr.io/0EPwJqxN5Q7

**Comment by JcolemanNR** [comment]
**Summary of Findings** 

My general observation is that the .NET runtime will create as many threads as the OS can spare in high throughput applications. This increase does occur more quickly with the .NET Agent attached to an application, but I did not observe radically different maximum thread counts with/without the agent attached over time. 

After reviewing multiple performance counters related to overall application performance/CPU usage, my assertion is that the root cause of this issue relates to lock contention introduced by the agent. Without the agent attached, I experienced a maximum of <100 contested locks per second in a high throughput application. With the agent attached, I experienced peaks of up to lock contention >30,000 per second. This metric has been also flagged as out of spec for high performance applications by Microsoft when customers have engaged them to troubleshoot application performance.

The agent makes heavy use of interlocked counters, and explicit locks to guard collections. If the lock contention could be reduced, I believe there would be less performance overhead.
